velox/functions/sparksql/aggregates/ModeAggregate.cpp

```diff
@@ -0,0 +1,585 @@
+      for (const auto& [key, count] : other) {
+        if (count.has_value()) {
+          values[key] += count.value();
+        }
+      }
+    }
+
+    bool writeIntermediateResult(out_type<IntermediateType>& out) {
+      out.reserve(values.size());
+      for (const auto& [value, count] : values) {
+        out.emplace(value, count);
+      }
+      return true;
+    }
+
+    bool writeFinalResult(out_type<OutputType>& out) {
+      auto iterator = std::max_element(
+          std::begin(values),
+          std::end(values),
+          [](const std::pair<T, int64_t>& p1, const std::pair<T, int64_t>& p2) {
+            return p1.second < p2.second;
+          });
+      out = iterator->first;
+      return true;
+    }
+
+    void destroy(HashStringAllocator* /*allocator*/) {
+      using TValues = decltype(values);
+      values.~TValues();
+    }
+  };
+};
+
+class StringModeAggregate {
+ public:
+  using InputType = Row<Varchar>;
+
+  using IntermediateType = Map<Varchar, int64_t>;
+
+  using OutputType = Varchar;
+
+  struct AccumulatorType {
+    using ValueMap = folly::F14FastMap<
+        StringView,
+        int64_t,
+        std::hash<StringView>,
+        std::equal_to<StringView>,
+        AlignedStlAllocator<std::pair<const StringView, int64_t>, 16>>;
+
+    // A map of unique StringViews pointing to storage managed by 'strings'.
+    ValueMap values;
+
+    // Stores unique non-null non-inline strings.
+    Strings strings;
+
+    explicit AccumulatorType(HashStringAllocator* allocator)
+        : values{AlignedStlAllocator<std::pair<const StringView, int64_t>, 16>(
+              allocator)} {}
+    static constexpr bool is_fixed_size_ = false;
+
+    void addInput(HashStringAllocator* allocator, arg_type<Varchar> data) {
+      values[store(data, allocator)]++;
+    }
+
+    void combine(
+        HashStringAllocator* allocator,
+        arg_type<IntermediateType> other) {
+      for (const auto& [key, count] : other) {
+        if (count.has_value()) {
+          values[store(key, allocator)] += count.value();
+        }
+      }
+    }
+
+    bool writeIntermediateResult(out_type<IntermediateType>& out) {
+      out.reserve(values.size());
+      for (const auto& [value, count] : values) {
+        out.add_item() = std::make_tuple(value, count);
+      }
+      return true;
+    }
+
+    bool writeFinalResult(out_type<OutputType>& out) {
+      auto iterator = std::max_element(
+          std::begin(values),
+          std::end(values),
+          [](const std::pair<StringView, int64_t>& p1,
+             const std::pair<StringView, int64_t>& p2) {
+            return p1.second < p2.second;
+          });
+      out = iterator->first;
+      return true;
+    }
+
+    void destroy(HashStringAllocator* allocator) {
+      strings.free(*allocator);
+      using TValues = decltype(values);
+      values.~TValues();
+    }
+
+   private:
+    StringView store(StringView value, HashStringAllocator* allocator) {
```

Comment:
Would you add some comment for this function?

Reply:
Added, thanks